### DISCUSSION TITLE
operator overloading

### DISCUSSION THREAD

**Discussion by felixroos** [tidalcycles/strudel#66] [Ideas]
One possible feature that we haven't really explored is operator overloading. It's also a decision of how much "strangeness", we let into strudel syntax, aka language strangeness budget, as @yaxu pointed out. It is generally possible to overload operators as part of the eval package, so the questions are:

- Should we even add overloaded operators?
- If yes, which ones? see https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators for a list of candidates
- Compatibility with tidal cannot really be met, maybe some common operators with haskell (if any) and python / vortex (at least >>>) can be found

**Comment by yaxu** [comment]
I think having some overloaded operators is inevitable. Just methods would work, but in a way a method is an overloading of the `.` operator, and I think only having that one operator is a bit too limiting. However, anything achievable with an overloaded operator will/should probably also be achievable with/be shorthand for a class method.

I think basic arithmetic like `+` `-` `*` `/` `%` `**` should keep its meaning and just be lifted into operating pattern values, where structure ('whole' timespans in event fragments) is preserved from the left. We currently have `add`, `sub`, `mul` and `div` methods for this, which should probably be renamed to something like `addLeft` etc, with `addRight` and `addBoth` etc for versions that preserve structure from left, right and both sides. I'm not sure if the bare `add` should then be an alias for `addLeft` like `+`, or `addBoth`.

Arrow-like operators `<` `>` `<=` `>=` `<<=` `>>=` and `>>>` have high value in e.g. manipulating time/chaining things together and should be spent wisely :) Likewise `&` `|` `^` `&&` `||` `??`. 

Could we potentially overload `in`?

**Comment by felixroos** [comment]
related: https://github.com/tidalcycles/strudel/discussions/82 https://github.com/tidalcycles/strudel/discussions/96
